[FileSystem] Move File Cache settings to GUI settings

thexai opened the pull request:
## Description
- Move File Cache settings to GUI settings (no functional changes).
- Provides compatibility for translate Bytes units in previous PR https://github.com/xbmc/xbmc/pull/24019

## Motivation and context
More easy cache settings for users and less headaches. Easier documentation using built-in GUI help strings.

Almost no functional changes, only some minor differences in File Cache defaults:

- Buffer Mode not changes default still `CACHE_BUFFER_MODE_NETWORK = 4`
~- Memory Size default increased a little from 20 MB to 32 MB (all current devices should capable handle it)~
~- Read Factor lowered from 4.0 to 1.5 reason: in some UHD streams (100 Mbit/s) * 4.0 is more than Wi-Fi  or USB connection can handle anyway and in some devices read at limit speeds also can cause CPU spikes. In other words: with big caches sizes is not necessary fill entire cache at fist second that playback starts but is better fill slowly when bitrate < avg. bitrate * 1.5~
- Chunk Size not changes default is 128 KB

Keep ALL  same defaults for now for smooth transition and low risk PR.

Thinking also in add a new chunk size setting only for local files but not in this PR yet.


## How has this been tested?
Runtime Windows x64 and Android (Shield)

## What is the effect on users?
No functional changes but users with personalized advancedsettings should re-configure File Cache in GUI (only a few clicks). The old settings in advancedsettings.xml will no longer have any effect.

## Screenshots (if appropriate):

![screenshot00000](https://github.com/xbmc/xbmc/assets/58434170/2aebf22d-1791-4f52-972b-cba34225914e)

![screenshot00001](https://github.com/xbmc/xbmc/assets/58434170/59e6ccc6-dd1a-43e9-a6cf-03582f4095c5)

![screenshot00002](https://github.com/xbmc/xbmc/assets/58434170/bd67a5ab-2fab-4122-a0d4-690bde91481a)

![screenshot00003](https://github.com/xbmc/xbmc/assets/58434170/25e28449-f9d5-40a7-ad93-cc5d44fb9e75)

![screenshot00004](https://github.com/xbmc/xbmc/assets/58434170/0c5e5b9f-29cb-4f26-93f3-48ccd28e96c3)

![screenshot00005](https://github.com/xbmc/xbmc/assets/58434170/5a73cd42-b2df-4d60-87aa-74071b5e8566)

![screenshot00006](https://github.com/xbmc/xbmc/assets/58434170/3b69dfa9-f39c-439b-8b63-5265dc5d84b0)

![screenshot00007](https://github.com/xbmc/xbmc/assets/58434170/8a863d6b-c47c-4397-be48-92d26c615d0e)

![screenshot00008](https://github.com/xbmc/xbmc/assets/58434170/7cae48b9-0ada-44a6-9aa8-a3df3db2c6fd)

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

thexai committed:
[FileSystem] Move File Cache settings to GUI settings

lrusak commented on xbmc/video/VideoThumbLoader.cpp:
```diff
@@ -483,10 +484,14 @@ std::string CVideoThumbLoader::GetLocalArt(const CFileItem &item, const std::str
+
+  const auto settings = CServiceBroker::GetSettingsComponent()->GetSettings();
+
+  const bool cache_all =
```
```suggestion
  const bool cacheAll =
```

thexai replied:
Done

lrusak reviewed (commented):
It might not be the best idea to change the default settings in the same commit. I would keep the defaults the same and change the defaults in a separate commit or an entirely separate PR.

enen92 commented on xbmc/filesystem/IFileTypes.h:
```diff
@@ -40,6 +40,13 @@ namespace XFILE
   static const unsigned int READ_REOPEN = 0x100;
 
+// file cache buffer modes
+  static constexpr int CACHE_BUFFER_MODE_INTERNET = 0;
```
Unsigned int?

thexai replied:
Not seems necessary because is not used as bitmask, only is compared with settings value which is also `int`:

`settings->GetInt(CSettings::SETTING_FILECACHE_BUFFERMODE) == CACHE_BUFFER_MODE_NETWORK`

enen92 commented on xbmc/settings/AdvancedSettings.cpp:
```diff
@@ -849,15 +841,6 @@ void CAdvancedSettings::ParseSettingsFile(const std::string &file)
     XMLUtils::GetString(pElement, "catrustfile", m_caTrustFile);
   }
 
-  pElement = pRootElement->FirstChildElement("cache");
```
It'd be great if instead of just removing this we could somehow migrate the settings, removing it completely by v22

thexai replied:
This is not trivial because not all possible values can be translated to new settings (only some fixed steps values).

It would be complicated to implement and in case of errors it could be worse. Additionally, some users have meaningless values such as Read Factor = 999 that comes from some add-ons and "wizards" advancedsettings generators :)

enen92 replied:
I am happy with the provided solution (log)

enen92 commented:
Looking at the screenshots I feel they are better located as a new group in the general settings than as a new entry in the menu. File cache is not exactly a service but something that applies to all of them

honest-mule commented:
Very well, this was much needed. Wanted to implement this myself but I'm very short on time these days.

Please keep up the good work!

garbear commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -22173,7 +22173,125 @@ msgctxt "#37057"
+#. Value of setting - KByte
+#: xbmc/settings/SevicesSettings.cpp
+msgctxt "#37121"
+msgid "KB"
```
Should these strings include the `{d}` placeholder for RTL languages?

thexai replied:
Done

jp12git commented:
💯 Fantastic! Don't need to mess around with adb to edit advancedsettings.xml file anymore

chewitt commented:
@enen92 @thexai It's not a "service" but it's associated to the SMB/NFS clients (which are also not services) so they need to kept together. Perhaps all of these should be moved to Settings > System?. I'd also suggest "Caching" for the menu item since if you configure the cache for all media it applies to internet streams which are not conventional files either. Single words normally translate better too.

thexai commented:
Addressed almost all commented:

- Renamed to "Caching" I agree that seems better name but not moved because also think is better keep together with NFS and SMB client settings (I proffer not move all to keep this PR simple and low risk for v21).
- Updated translation strings for RTL languages (`{0:d}` placeholder).
- Keep the same defaults for now as requested @lrusak (this also helps to low risk PR and smooth transition).
- Added a new Log message as not lose the possibility of see the cache settings from the debug logs and also to avoid confusions due to the change and old values still in advancedsettings.xml

This is at app start only:
```
2023-11-02 17:11:10.659 T:9524     info <general>: New cache GUI settings (replacement of cache advancedsettings.xml) are:
                                                    - Buffer Mode: 4
                                                    - Memory Size: 20 MB
                                                    - Read Factor: 4.00 x
                                                    - Chunk Size : 131072 bytes
```

arnova commented:
Looks good. Only thing I like to mention: Ideally the buffer-mode should be set on a per-source basis that would probably eliminate a lot of issue reports... Probably something to keep in mind while implementing this.

thexai commented:
> Ideally the buffer-mode should be set on a per-source basis that would probably eliminate a lot of issue reports... Probably something to keep in mind while implementing this.

Maybe in a future PR but the goal of this was move settings from advancedsettings to GUI Settings without change any functionality.

thexai commented:
@jjd-uk, Can you review settings xml changes and strings.po changes?

chewitt commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -22173,7 +22173,125 @@ msgctxt "#37057"
+#. Description of category #37101 "Caching"
+#: system/settings/settings.xml
+msgctxt "#37102"
+msgid "This category contains the settings for configure Caching of Internet streams, network files and local files."
```
This category contains settings that configure caching of local and network files, and internet streams."

chewitt commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -22173,7 +22173,125 @@ msgctxt "#37057"
+#. Description of setting #37103 "Buffer Mode"
+#: system/settings/settings.xml
+msgctxt "#37104"
+msgid "Determines what type of contents are buffered."
```
"Configure the media content to buffer."

chewitt commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -22173,7 +22173,125 @@ msgctxt "#37057"
+#. Description of setting #37105 "Memory Size"
+#: system/settings/settings.xml
+msgctxt "#37106"
+msgid "Number of bytes used for buffering streams in memory.[CR]When set to 0 the cache will be written to disk instead of RAM (not recommended for SSD/Flash storage)."
```
"The size of the memory buffer in bytes. Setting zero (0) forces buffering to disk, which is not recommend on flash storage devices (eMMC, SD cards, SSD)."

thexai commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -22173,7 +22173,125 @@ msgctxt "#37057"
+#. Description of setting #37107 "Read Factor"
+#: system/settings/settings.xml
+msgctxt "#37108"
+msgid "This factor determines the max read rate in terms of Read Factor x avg. bitrate of a video file.[CR]Chosen value should be realistic for the bitrate of the content and the capacity of the connection e.g. if the max UHD Blu-Ray bitrate is 140 Mbit/s and the connection is 250 Mbit/s, the maximum Read Factor is 250/140 = 1.78x"
```
I'm not sure of this one:

> Increasing the multiple forces Kodi to read more data into the cache with each read request

Amount of data read in each read request depends on Chunk Size

> Large multiples will delay playback start

This is not true from https://github.com/xbmc/xbmc/pull/23760 --> playback start depends only on Video Player queue time (> 8 seconds) In fact playback start when queue is full and cache starts filling (first byte received). See https://github.com/xbmc/xbmc/pull/23760#issuecomment-1722389237

Maybe:
"The read factor determines cache fill rate in terms of avg. bitrate of stream x Read Factor. Increasing this multiple, cache fills faster but more bandwidth is consumed. Use large multiples may cause more CPU usage on some devices, saturate the connection and worsen performance."

honest-mule replied:
The read-factor needs to be abandoned, of course not in the scope of this PR but the point is a thorough discussion should be done about the utility and application of this variable.

The code where rate is being manipulated using the read-factor is not too well understood by anybody and this is what I could gather from looking into history of the file `FileCache.cpp`. Perhaps @arnova has that particular knowledge about how that line of code is doing it's magic? If so, it would be better to make the line of code at least comprehensible in such a manner so that it's utility can be recognized in an obvious way.

In all of my testing removing that line of code had little impact on how soon the cache was being filled. Why? Because the memory cache is always limited and once the cache is full no further writes are possible which naturally rate-limits Kodi from reading more from the stream.

The other scenario where rate-limiting might be making sense to everyone that cares is this: let's suppose someone has set ~4GB or similar memory cache for Kodi -- if someone has done this indeed, we can freely assume they've the know-how to understand their internet will be utilized greedily to fill that kind of cache size, in fact. that's the whole point of having a big cache, the user wants Kodi to fill it greedily for their smooth playback needs. Had their internet were that good where a small cache would get filled instantly, they would've never made or at all needed to make the cache size bigger.

More than likely this concept of rate-limiting has been inspired from some internet video providers which provide their own app to the users, like Netflix. Now Netflix has the responsibility and drive to keep their servers available for all of their customers, so obviously they've rate-limiting baked in their app. However Kodi has no such responsibility of maintaining or caring for any such servers, it can indeed read all that the user wants it to read.

@fritsch I hope you'll read this. Would it make sense to make a PR for this?

fritsch replied:
Sense for this PR: Get it in as fast as possible as is ... else bikeshedding will kill it :-)

And on topic: yes readrate is the next big thingy. It has different aspects depending on buffer is empty (filling it fast), but a flaw for filling a buffer back up, when readrate x factor does not fit in. This needs strong rework. I even think that this factor should not be exposed at all, but a better algorithm should make sure that the cache is always filled to a certain degree (PID regulator approach), so that we don't run in too much spikes.

thexai replied:
> Sense for this PR: Get it in as fast as possible as is ... else bikeshedding will kill it :-)

I'm waiting to someone approve it to merge 😉

fritsch replied:
Yeah - I would do that right away, but I was not involved in the delta-discussions, which is why I would prefer if someone of those colleagues would give their ack.

arnova replied:
> The read-factor needs to be abandoned, of course not in the scope of this PR but the point is a thorough discussion should be done about the utility and application of this variable.
> 
> The code where rate is being manipulated using the read-factor is not too well understood by anybody and this is what I could gather from looking into history of the file `FileCache.cpp`. Perhaps @arnova has that particular knowledge about how that line of code is doing it's magic? If so, it would be better to make the line of code at least comprehensible in such a manner so that it's utility can be recognized in an obvious way.
> 
I'm not sure what "line of code" you're referring to?

thexai replied:
> The read-factor needs to be abandoned, of course not in the scope of this PR but the point is a thorough discussion should be done about the utility and application of this variable.

> The code where rate is being manipulated using the read-factor is not too well understood by anybody and this is what I could gather from looking into history of the file

> In all of my testing removing that line of code had little impact on how soon the cache was being filled

Seems you suggest all code is not explained/understood (or you not understood) should be removed :)

This is only an example of how works Read Factor: as mentioned determines cache fill rate... 

It's just a slightly extreme example with very large cache sizes and "low" read factors.

**Read Factor = 4.0x** (took 34s to fill 191 MB of cache)

https://github.com/xbmc/xbmc/assets/58434170/feed767e-672f-48b2-b439-22d29b93b30e


**Read Factor = 1.5x** (took 2m 39s to fill 191 MB of cache) recording split in two parts in two parts

https://github.com/xbmc/xbmc/assets/58434170/c4c5fbd4-f116-487a-abfc-7d558c364dc9

https://github.com/xbmc/xbmc/assets/58434170/291d4864-d780-477d-b0a1-f2113437e748

thexai replied:
Kodi is a player, not a downloader. The mission of a downloader may be to download at the highest possible speed. A player must "download" at just the right speed to play smoothly. We should prioritize the playback thread more than the download thread and of course should not waste/use more CPU than necessary to download. In some devices, while playing 4K Dolby Vision, the CPU can be at 60 -70% if an Ethernet connection is also managed, downloading at 800 Mbps, CPU peaks can be generated that end up causing frame drops/skips. Limiting the download speed to only 150 - 200 Mbps (for a 100 Mbps stream) avoids unnecessary CPU spikes and is not required cache 100% filled to start playback.

thexai replied:
I have very little free time and this is all off topic because this PR is not intended to change the functionality. I'm not going to waste any more time on this discussion.

honest-mule replied:
Would've been better if you had not jumped in at all with your needless diatribe about how, "Kodi isn't a downloader" shtick. I was here looking for an explanation/reasoning, not from you but from the devs that maybe-probably understand that code section better than anybody else.

But seems like you've wasted not only "the very little free time" of yours, but very valuable time of mine as well, and almost everyone else's time who will end up reading this garbage because you couldn't control yourself from polluting the thread with conjectures.

I've seen this behavior from you in the past and I highly recommend that you start developing the habit of comprehensively reading & understanding a post's content thoroughly before hitting that reply button.

fritsch commented:
Sense for this PR: Get it in as fast as possible as is ... else bikeshedding will kill it :-)

arnova commented:
It's a shame this was closed (and you deleted the branch as well). I considered the debate completely beyond the context of this PR.

wsnipex commented:
please consider reopening

fritsch commented:
@thexai would be great if you could reopen. This change is awesome and helps a lot for user's boilerplates. We have plenty of room for additional bike shedding in a non-related PR of someone else :-)

jjd-uk reviewed (approved):
Approval for the xml settings changes only. Can't say anything at all about the C++ side.

While I'm not 100% convinced we need a new settings category for caching,  I do not see any reason to hold this up any further, as there seems to have been plenty of bike shedding. I think perhaps the whole settings layout needs to be reviewed for v22 as there's been lots of recent additions/changes.

thexai commented:
Was missing unregister of new settings fillers

thexai commented:
Jenkins build this please

pannal commented:
With reference [to this](https://github.com/xbmc/xbmc/pull/23760#issuecomment-1814383711), I'd like to see the advancedsettings buffer/readsize values retained, honored and maybe setting the default in the GUI?

This way, users don't need to reconfigure the cache settings when switching to Omega and PlexMod4Kodi can still manage the buffer/readsize settings for users.

pannal commented:
I tend to agree, but there're multiple reasons for why it should continue to be possible:
* it's been possible "forever":
  * `advancedsettings.xml` is in the user directory, thus writable by plugins
  * `XMLRPC.SetSettingValue` is an official API endpoint
* most users of Plex4Kodi and PlexMod4Kodi use Kodi _only_ for these addons, mostly on streaming boxes/sticks. I've significantly lowered the entrypoint by providing proper buffer settings and the ability to write them directly.

The Plex4Kodi derivatives are much more capable than the official Plex apps, especially with passthrough and bad connectivity situations.

> That is to say, an add-on should not be able to change/break the proper parameters to playback local files or SMB / NFS network files that not using that add-on.
Correct. PM4K parses existing `advancedsettings.xml` and only touches the buffer size and readfactor if requested by the user. It doesn't touch the buffer mode.

Nevertheless:
I won't argue against this PR or its changes. I've implemented a "fall-forward" and PM4K will use `SetSettingValue` instead of writing to `advancedsettings.xml` when running on a compatible Kodi version.

thexai commented:
> I won't argue against this PR or its changes. I've implemented a "fall-forward" and PM4K will use SetSettingValue instead of writing to advancedsettings.xml when running on a compatible Kodi version.

Correct. This PR does not change any functionality, it just moves some settings from one place to another...

The proper solution is fix `Player.ProgressCache` in case it doesn't work well (which is likely).
